src/bitcoin.rs

```diff
@@ -0,0 +1,120 @@
+use crate::dai;
+use crate::dai::ATTOS_IN_DAI_EXP;
+use crate::publish::WorthIn;
+use crate::rate::Rate;
+use num::pow::Pow;
+use num::BigUint;
+
+pub const SATS_IN_BITCOIN_EXP: u16 = 8;
+
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, PartialEq, Eq)]
+pub struct Amount(::bitcoin::Amount);
+
+impl Amount {
+    pub fn from_btc(btc: f64) -> anyhow::Result<Amount> {
+        Ok(Amount(::bitcoin::Amount::from_btc(btc)?))
+    }
+
+    pub fn from_sat(sat: u64) -> Self {
+        Amount(::bitcoin::Amount::from_sat(sat))
+    }
+
+    pub fn as_sat(self) -> u64 {
+        self.0.as_sat()
+    }
+
+    pub fn as_btc(self) -> f64 {
+        self.0.as_btc()
+    }
+}
+
+// The rate input is for bitcoin to dai but we applied to satoshis so we need to:
+// - divide to get bitcoins (8)
+// - divide to adjust for rate (9)
+// - multiply to get attodai (18)
+// = 1
+const ADJUSTEMENT_EXP: u16 = ATTOS_IN_DAI_EXP - SATS_IN_BITCOIN_EXP - Rate::PRECISION;
+
+impl WorthIn<dai::Amount> for Amount {
+    fn worth_in(&self, btc_to_dai_rate: Rate) -> anyhow::Result<dai::Amount> {
+        // Get the integer part of the rate
+        let uint_rate = btc_to_dai_rate.integer();
+
+        // Apply the rate
+        let worth = uint_rate * self.as_sat();
+
+        let atto_dai = worth * BigUint::from(10u16).pow(BigUint::from(ADJUSTEMENT_EXP));
+
+        Ok(dai::Amount::from_atto(atto_dai))
+    }
+}
```

Comment:
If you don't necessarily need the trait (where are we abstracting over it?), then you could remove the `Result` from this one :)

Reply:
True, we were abstracting but it went away.